### PR TITLE
Two fixed point theorems

### PR DESCRIPTION
**Pull request by fredrik-bakke**
Formalizations of the Knaster–Tarski fixed point theorem and Kleene's fixed point theorem, in addition to the necessary infrastructure.

~Merge after #1225.~

### CODE REVIEW

File: src/order-theory/suplattices.lagda.md [hunk 1]
```diff
@@ -119,10 +126,24 @@ module _
...
       ( poset-Suplattice)
       ( is-suplattice-Suplattice)
+
+  leq-sup-Suplattice :
+    {I : UU l3} (x : I → type-Suplattice) (i : I) →
+    leq-Suplattice (x i) (sup-Suplattice x)
+  leq-sup-Suplattice x =
+    backward-implication
+      ( is-least-upper-bound-sup-Suplattice x (sup-Suplattice x))
+      ( refl-leq-Suplattice (sup-Suplattice x))
```
Comment (EgbertRijke) [inline src/order-theory/suplattices.lagda.md:144]: Wouldn't it be nicer to formalize this using the existing concept of upper bounds and follow the existing naming scheme?

```suggestion
  is-upper-bound-family-of-elements-sup-Suplattice :
    {I : UU l3} (x : I → type-Suplattice)
    is-upper-bound-family-of-elements-Poset
      ( poset-Suplattice)
      ( x)
      ( sup-Suplattice x)
  leq-sup-Suplattice x =
    backward-implication
      ( is-least-upper-bound-sup-Suplattice x (sup-Suplattice x))
      ( refl-leq-Suplattice (sup-Suplattice x))
```

File: src/order-theory/supremum-preserving-maps-posets.lagda.md [hunk 2]
```diff
@@ -0,0 +1,333 @@
+# Supremum preserving maps on posets
```
Comment (EgbertRijke) [inline src/order-theory/supremum-preserving-maps-posets.lagda.md:1]:
```suggestion
# Supremum preserving maps between posets
```

File: src/order-theory/supremum-preserving-maps-posets.lagda.md [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+  abstract
+    preserves-joins-preserves-suprema-Poset :
+      {f : type-Poset P → type-Poset Q} →
+      preserves-suprema-Poset P Q l5 f →
+      preserves-joins-Poset P Q f
+    preserves-joins-preserves-suprema-Poset {f} H x y s =
+      is-least-binary-upper-bound-has-least-upper-bound-family-of-elements-Poset
+        ( Q)
+        ( rec-bool (f x) (f y))
+        ( f (pr1 s) , ？) -- TODO
```
Comment (EgbertRijke) [inline src/order-theory/supremum-preserving-maps-posets.lagda.md:162]: A todo is left open here

File: src/order-theory/supremum-preserving-maps-posets.lagda.md [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+```
+
+### The identity supremum preserving map
+
+```agda
+module _
+  {l1 l2 : Level} (P : Poset l1 l2)
+  where
+
+  preserves-suprema-id-Poset :
```
Comment (EgbertRijke) [inline src/order-theory/supremum-preserving-maps-posets.lagda.md:248]:
```suggestion
  preserves-suprema-id-supremum-preserving-hom-Poset :
```

File: src/order-theory/supremum-preserving-maps-posets.lagda.md [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+
+### Composing supremum preserving maps
+
+```agda
+module _
+  {l1 l2 l3 l4 l5 l6 l7 : Level}
+  (P : Poset l1 l2) (Q : Poset l3 l4) (R : Poset l5 l6)
+  where
+
+  preserves-suprema-comp-Poset :
```
Comment (EgbertRijke) [inline src/order-theory/supremum-preserving-maps-posets.lagda.md:264]:
```suggestion
  preserves-suprema-comp-supremum-preserving-hom-Poset :
```

File: src/order-theory/inflattices.lagda.md [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+      ( poset-Inflattice)
+      ( is-inflattice-Inflattice)
+
+  leq-inf-Inflattice :
+    {I : UU l3} (x : I → type-Inflattice) (i : I) →
+    leq-Inflattice (inf-Inflattice x) (x i)
+  leq-inf-Inflattice x =
+    backward-implication
+      ( is-greatest-lower-bound-inf-Inflattice x (inf-Inflattice x))
+      ( refl-leq-Inflattice (inf-Inflattice x))
```
Comment (EgbertRijke) [inline src/order-theory/inflattices.lagda.md:141]: Formalize this as `is-lower-bound-family-of-elements-inf-Inflattice` using the existing definition of lower bounds?

File: src/order-theory/chains-posets.lagda.md [hunk 7]
```diff
@@ -91,7 +98,65 @@ module _
     is-prop-inclusion-chain-Preorder (preorder-Poset X) C D
 ```
 
+## Properties
+
+### Chains in a poset from order preserving maps from a total order
```
Comment (EgbertRijke) [inline src/order-theory/chains-posets.lagda.md:103]: Took me a couple of reads to parse.

```suggestion
### Any order preserving maps from a total order into a poset induces a chain
```

**Review by EgbertRijke — COMMENTED** [review]
I'm very excited to merge this pull request into the library! Very well done.

I had some stylistic remarks and some questions about your naming.

File: src/domain-theory/directed-complete-posets.lagda.md [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+open import order-theory.least-upper-bounds-posets
+open import order-theory.posets
+```
+
+</details>
+
+## Idea
+
+A
+{{#concept "directed complete poset" WD="complete partial order" WDID=Q3082805 Agda=Directed-Complete-Poset}}
```
[the remark this answers is not in the document]
Comment (fredrik-bakke) [reply src/domain-theory/directed-complete-posets.lagda.md:30]: The Wikipedia article begins:

> In [mathematics](https://en.wikipedia.org/wiki/Mathematics), the phrase complete partial order is variously used to refer to at least three similar, but distinct, classes of [partially ordered sets](https://en.wikipedia.org/wiki/Partially_ordered_set), [...]

It goes on to describe DCPOs among others, so Wikipedia itself refers to DCPOs as complete partial orders.

[the remark this answers is not in the document]
Comment (EgbertRijke) [reply src/domain-theory/directed-complete-posets.lagda.md:30]: Ok, thanks for clarifying! Given that Wikidata conflates these notions it is indeed appropriate to use these WD identifiers. Thanks!

File: src/domain-theory/kleenes-fixed-point-theorem-omega-complete-posets.lagda.md [hunk 9]
```diff
@@ -0,0 +1,304 @@
...
+
+## Idea
+
+{{#concept "Kleene's fixed point theorem" Disambiguation="ω-complete posets" WD="Kleene fixed-point theorem" WDID=Q3527263}}
+states that given an
+[ω-continuous](domain-theory.omega-continuous-maps-omega-complete-posets.md)
+endomap `f : 𝒜 → 𝒜` on an
+[ω-complete poset](domain-theory.omega-complete-posets.md) `𝒜`, then for every
+`x ∈ 𝒜` such that `x ≤ f x`, the ω-transfinite application of `f` to `x`,
+`f^ω(x)`, given that it exists, is a
```
[the remark this answers is not in the document]
Comment (fredrik-bakke) [reply src/domain-theory/kleenes-fixed-point-theorem-omega-complete-posets.lagda.md:51]: It is hypothesized that it exists. is the rephrasing "when it exists" sufficiently clear?

[the remark this answers is not in the document]
Comment (EgbertRijke) [reply src/domain-theory/kleenes-fixed-point-theorem-omega-complete-posets.lagda.md:51]: Yes, or "if it exists"

File: src/domain-theory/omega-continuous-maps-omega-complete-posets.lagda.md [hunk 10]
```diff
@@ -0,0 +1,398 @@
...
+
+```text
+  x₀ ≤ x₁ ≤ x₂ ≤ ­… ≤ xₙ ≤ xₙ₊₁ ≤ … ≤ xω
+```
+
+to the supremum of the image of the ascending chain
+
+```text
+  f x₀ ≤ f x₁ ≤ f x₂ ≤ ­… ≤ f xₙ ≤ f xₙ₊₁ ≤ … ≤ f xω.
+```
```
[the remark this answers is not in the document]
Comment (fredrik-bakke) [reply src/domain-theory/omega-continuous-maps-omega-complete-posets.lagda.md:61]: That's a good point, thanks for pointing it out!

File: src/domain-theory/kleenes-fixed-point-theorem-omega-complete-posets.lagda.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (fredrik-bakke) [reply src/domain-theory/kleenes-fixed-point-theorem-omega-complete-posets.lagda.md:51]: Oh, haha, I realize why there's confusion here. Indeed, the limit exists when the poset is ω-complete.

[the remark this answers is not in the document]
Comment (fredrik-bakke) [reply src/domain-theory/kleenes-fixed-point-theorem-omega-complete-posets.lagda.md:51]: This was copy-pasted from the non-complete case. Sorry about that

File: src/foundation/surjective-maps.lagda.md [hunk 11]
```diff
@@ -830,6 +837,26 @@ module _
...
+  where
+
+  is-inhabited-is-surjective :
+    {f : A → B} → is-surjective f → is-inhabited B → is-inhabited A
+  is-inhabited-is-surjective F =
+    rec-trunc-Prop
+      ( is-inhabited-Prop A)
+      ( rec-trunc-Prop (is-inhabited-Prop A) (unit-trunc-Prop ∘ pr1) ∘ F)
+
+  is-inhabited-surjects-onto :
```
[the remark this answers is not in the document]
Comment (fredrik-bakke) [reply src/foundation/surjective-maps.lagda.md:854]: The proposed name suggests to me that the type of surjections is inhabited, not that the domain is. Perhaps there is a less ambiguous name we can give this entry?

File: src/order-theory/supremum-preserving-maps-posets.lagda.md [hunk 12]
```diff
@@ -0,0 +1,333 @@
...
+      preserves-suprema-Poset P Q l5 f →
+      preserves-joins-Poset P Q f
+    preserves-joins-preserves-suprema-Poset {f} H x y s =
+      is-least-binary-upper-bound-has-least-upper-bound-family-of-elements-Poset
+        ( Q)
+        ( rec-bool (f x) (f y))
+        ( f (pr1 s) , ？) -- TODO
+```
+
+> From this property the next would be a simple corollary.
```
[the remark this answers is not in the document]
Comment (fredrik-bakke) [reply src/order-theory/supremum-preserving-maps-posets.lagda.md:165]: it is not formalized that way since the above code is unfinished 😶‍🌫️

File: src/foundation/surjective-maps.lagda.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (fredrik-bakke) [reply src/foundation/surjective-maps.lagda.md:854]: how about `is-inhabited-domain-surjection`?

File: src/order-theory/supremum-preserving-maps-posets.lagda.md (hunk 3 above)
Comment (fredrik-bakke) [reply src/order-theory/supremum-preserving-maps-posets.lagda.md:162]: Yes... do I _have_ to finish it? 🥺

File: src/order-theory/resizing-preorders.lagda.md [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+  is-left-inverse-hom-inv-resize-type-Preorder :
+    htpy-hom-Preorder (resize-type-Preorder P e) (resize-type-Preorder P e)
+      ( comp-hom-Preorder
+        ( resize-type-Preorder P e)
+        ( P)
+        ( resize-type-Preorder P e)
+        ( hom-inv-resize-type-Preorder)
+        ( hom-resize-type-Preorder))
+      ( id-hom-Preorder (resize-type-Preorder P e))
+  is-left-inverse-hom-inv-resize-type-Preorder = is-retraction-map-inv-equiv e
```
[the remark this answers is not in the document]
Comment (fredrik-bakke) [reply src/order-theory/resizing-preorders.lagda.md:92]: That concept is not formalized yet, which is why this was left like this. I should probably add that 😅

File: src/order-theory/join-preserving-maps-posets.lagda.md [hunk 14]
```diff
@@ -0,0 +1,282 @@
...
+```
+
+### The identity join preserving map
+
+```agda
+module _
+  {l1 l2 : Level} (P : Poset l1 l2)
+  where
+
+  preserves-joins-id-Poset :
```
[the remark this answers is not in the document]
Comment (fredrik-bakke) [reply src/order-theory/join-preserving-maps-posets.lagda.md:196]: following your other suggestions this should be `preserves-joins-id-hom-Poset`, it seems

File: src/foundation/surjective-maps.lagda.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (EgbertRijke) [reply src/foundation/surjective-maps.lagda.md]: We do have a naming scheme that allows us to write stuff like `is-inhabited-equiv`, `is-finite-equiv`, `is-trunc-equiv` and so on, where the intended reading is to conclude something from having an equivalence. 

In the library we also currently have `is-inhabited-is-contr`, which according to one reading could say that the type `is-contr A` is inhabited, but the meaning here is to conclude `is-inhabited` from `is-contr`.

I could be ok with `is-inhabited-domain-surjection` but then I still wonder if by doing so we start to undermine our current naming practice, and I don't really want to get into a situation where we have to start renaming things across the whole library.

I'll let you make the call.

File: src/order-theory/supremum-preserving-maps-posets.lagda.md (hunk 3 above)
Comment (EgbertRijke) [reply src/order-theory/supremum-preserving-maps-posets.lagda.md:166]: Perhaps remove the unfinished code block?

File: src/order-theory/resizing-preorders.lagda.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (EgbertRijke) [reply src/order-theory/resizing-preorders.lagda.md]: Or just constructing the equivalence would be fine.

[the remark this answers is not in the document]
Comment (fredrik-bakke) [reply src/order-theory/resizing-preorders.lagda.md]: Wouldn't "just the equivalence" be the underlying equivalence of the smallness witness?

File: src/domain-theory/directed-complete-posets.lagda.md [hunk 15]
```diff
@@ -34,15 +34,15 @@ is a [poset](order-theory.posets.md) such that all
 
 ## Definitions
 
-### The predicate on posets of being a directed complete poset
+### The predicate on posets of being a directed complete
```
Comment (EgbertRijke) [inline src/domain-theory/directed-complete-posets.lagda.md:37]: indefinite article

File: src/domain-theory/omega-complete-posets.lagda.md [hunk 16]
```diff
@@ -45,15 +45,15 @@ in `P`, i.e., ascending chain indexed by the
 
 ## Definitions
 
-### The predicate on posets of being an ω-complete poset
+### The predicate on posets of being a ω-complete
```
Comment (EgbertRijke) [inline src/domain-theory/omega-complete-posets.lagda.md:48]: typo: this indefinite article shouldn't be there

File: src/order-theory/inflattices.lagda.md [hunk 17]
```diff
@@ -15,14 +15,16 @@ open import foundation.sets
...
 open import order-theory.posets
 ```
 
 </details>
 
 ## Idea
 
-An `l`-{{#concept "inflattice"  Agda=Inflattice}} is a
+Consider a [universe level](foundation.universe-levels.md) `l`. An
+`l`-{{#concept "inflattice"  Agda=Inflattice}} is a
```
Comment (EgbertRijke) [inline src/order-theory/inflattices.lagda.md:27]: Typo: Universe levels at the end of this paragraph shouldn't be hyperlinked anymore

File: src/order-theory/resizing-preorders.lagda.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (fredrik-bakke) [reply src/order-theory/resizing-preorders.lagda.md]: Oh, I figured it out. Here the parametric definition of the large precategory of posets actually comes in handy :)

[the remark this answers is not in the document]
Comment (fredrik-bakke) [reply src/order-theory/resizing-preorders.lagda.md]: Well, uh, I can formalize the equivalence of posets this way at least. However, preorders and order preserving maps don't define a set-level category, and I would prefer not getting into wild category theory territory just to define this equivalence for preorders. Is it okay if I leave it out?

[the remark this answers is not in the document]
Comment (EgbertRijke) [reply src/order-theory/resizing-preorders.lagda.md]: Definitely. Let's focus on getting this merged

File: src/elementary-number-theory/decidable-total-order-natural-numbers.lagda.md [hunk 18]
```diff
@@ -41,3 +48,50 @@ pr1 ℕ-Decidable-Total-Order = ℕ-Poset
 pr1 (pr2 ℕ-Decidable-Total-Order) = is-total-leq-ℕ
 pr2 (pr2 ℕ-Decidable-Total-Order) = is-decidable-leq-ℕ
 ```
+
+## Properties
+
+### Defining order preserving maps out of the natural numbers by induction
```
[the remark this answers is not in the document]
Comment (fredrik-bakke) [reply src/elementary-number-theory/decidable-total-order-natural-numbers.lagda.md:54]: what would you rather call it? Something like the following?

```suggestion
### Defining order preserving maps out of the natural numbers successorwise
```

[the remark this answers is not in the document]
Comment (EgbertRijke) [reply src/elementary-number-theory/decidable-total-order-natural-numbers.lagda.md:54]: The map is not really being defined, because it is assumed. It's just the proof of it being order preserving that is handled here by induction.

[the remark this answers is not in the document]
Comment (fredrik-bakke) [reply src/elementary-number-theory/decidable-total-order-natural-numbers.lagda.md]: Ah gotcha. Let me know if the new formulation should be revised further

**Comment by fredrik-bakke** [comment]
I believe I've addressed all of your comments now; let me know if there is anything else. Thank you for the helpful comments!

File: src/domain-theory/directed-families-posets.lagda.md [hunk 19]
```diff
@@ -40,7 +40,7 @@ both `x i ≤ x k` and `x j ≤ x k` hold.
 
 ## Definitions
 
-### The predicate on a family of elements of a poset of being directed
+### The predicate on a family of elements in a poset of being directed
```
Comment (EgbertRijke) [inline src/domain-theory/directed-families-posets.lagda.md:43]: I'm pretty sure that "element of" is perfectly fine and perhaps even more common, but I won't make you undo this change. I'm just not sure if I want to have a policy of having to say "element in".
